Make use of xtables.lock for parallel iptables usage

aanm opened the pull request:
Adding xtables.lock usage into cilium. Review per commit basis.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8701)
<!-- Reviewable:end -->

aanm committed:
examples/kubernetes: mount xtables.lock

As Cilium access iptables it should also mount the xtables.lock
file into iptables.

Signed-off-by: André Martins <andre@cilium.io>

brb commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -286,6 +290,7 @@ func (m *IptablesManager) RemoveRules() {
 
 func ingressProxyRule(cmd, l4Match, markMatch, mark, port, name string) []string {
 	ret := []string{
+		"-w",
```
`-w` without a duration will make it block forever if someone holds the lock indefinitely. In this case, it makes sense to log an error and crash as otherwise it will be difficult to debug why cilium-agent doesn't make a progress during its initialization.

aanm replied:
@brb what do you thing is the right timeout here? 60 seconds?

brb replied:
Kubernetes uses 5sec (https://github.com/kubernetes/kubernetes/blob/master/pkg/util/iptables/iptables.go#L130). We could use this value as well.

Keep in mind that not all iptables versions support `-w`, so we need to do the feature detection.

aanm committed:
use iptables-manager to manage iptables executions

This will make it easier to run iptables commands based on iptables
version installed in the system.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
datapath/iptables: wait until acquisition xtables lock is done

This guarantees iptables interactions are not run concurretly with other
programs, e.g. kube-proxy.

More info: https://git.netfilter.org/iptables/commit/?id=93587a04d0f2511e108bbc4d87a8b9d28a5c5dd8

Signed-off-by: André Martins <andre@cilium.io>

brb commented on daemon/daemon.go:
```diff
@@ -683,8 +686,13 @@ func createPrefixLengthCounter() *counter.PrefixLengthCounter {
 
+type rulesManager interface {
+	RemoveRules()
+	InstallRules(ifName string) error
```
Not related to your changes, but can we rename `ifName` to `hostDevice`?

brb commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -72,13 +99,13 @@ func getFeedRule(name, args string) []string {
 	var err error
 	if option.Config.EnableIPv4 {
-		err = runProg("iptables", []string{"-t", c.table, "-N", c.name}, false)
+		err = runProg("iptables", append(waitArgs, "-t", c.table, "-N", c.name), false)
```
Instead of passing `waitArgs` to each occurrence of `runProg`, we could refactor each function which calls it in a way that a function would return list of rules, and then a higher-level function would invoke `iptables` with `waitArgs`.

brb reviewed (commented):
LGTM. However, I'd probably prefer to see the suggested refactoring in this PR to avoid passing `waitArgs` for each occurrence of `runProg`.
